packages/falcon-ecommerce-uikit/src/Search/SearchProvider.tsx

```diff
@@ -0,0 +1,172 @@
+import { searchStateToURL } from './searchStateToURL';
+import { SearchContext } from './SearchContext';
+
+export const SORT_ORDERS_QUERY = gql`
+  query SortOrdersQuery {
+    sortOrders @client
+  }
+`;
+
+interface SearchProviderProps extends RouteComponentProps {
+  searchStateToURL?(state: SearchState): string;
+  searchStateFromURL?(url: string): SearchState;
+  sortOrders: SortOrder[];
+}
+
+class SearchProviderImpl extends React.Component<SearchProviderProps, SearchState> {
+  static defaultProps = {
+    searchStateFromURL,
+    searchStateToURL,
+    sortOrders: []
+  };
+
+  constructor(props: SearchProviderProps) {
+    super(props);
+    this.state = this.getStateFromURL(props.location);
+    this.historyUnlisten = () => {};
+  }
+
+  componentDidMount() {
+    this.historyUnlisten = this.props.history.listen(this.restoreStateFromURL);
+  }
+
+  componentWillUnmount() {
+    this.historyUnlisten();
+  }
+
+  getStateFromURL(location: any) {
+    const state = this.props.searchStateFromURL!(location.search);
+    // if sort order was passed in url then make sure that it's available
+    if (state.sort) {
+      const sortOrderItem = this.getFullSortOrderDefinition(state.sort);
+      if (sortOrderItem) {
+        state.sort = sortOrderItem;
+      } else {
+        delete state.sort;
+      }
+    }
+
+    return state;
+  }
+
+  getFullSortOrderDefinition(sort: SortOrder) {
+    return this.props.sortOrders.find(
+      (item: SortOrder) => item.field === sort.field && item.direction === sort.direction
+    );
+  }
+
+  setFilter = (field: string, value: string[], operator = 'eq' as FilterOperator) => {
+    const filters = this.state.filters ? [...this.state.filters] : [];
+    let filter = filters.find(item => item.field === field);
+
+    if (!filter) {
+      filter = {
+        operator,
+        field,
+        value
+      };
+      filters.push(filter);
+    } else {
+      filter.operator = operator;
+      filter.value = value;
+    }
+
+    this.updateURL({
+      ...this.state,
+      filters
+    });
+  };
+
+  setSortOrder = (sort: SortOrder) => {
+    const sortItem = this.getFullSortOrderDefinition(sort);
+    if (!sortItem) {
+      throw new Error(
+        'Sort order value passed to SearchProvider.setSortOrder() does not match any of available sort orders'
+      );
+    }
+
+    this.updateURL({
+      ...this.state,
+      sort: sortItem
+    });
+  };
+
+  setPagination = (pagination: PaginationInput) => {
+    this.updateURL({
+      ...this.state,
+      pagination
+    });
+  };
+
+  setQuery = (query: string) => {
+    this.updateURL({
+      ...this.state,
+      query
+    });
+  };
+
+  removeFilter = (field: string) => {
+    if (!this.state.filters) {
+      return;
+    }
+
+    const filters = this.state.filters.filter(filter => filter.field !== field);
+
+    this.updateURL({
+      ...this.state,
+      filters
+    });
+  };
+
+  private historyUnlisten: Function;
```

Comment:
maybe we can merge this declaration with default value initialization from [ctor](https://github.com/deity-io/falcon/pull/309/files#diff-269cc7cda5d1b5b4e58c32320972d090R32)?

```suggestion
  private historyUnlisten = () => {};
```

Reply:
True, done :)